carbonmark/components/pages/Project/Retire/RetireForm.tsx

```diff
@@ -0,0 +1,272 @@
+    setInputValues(null);
+    setAllowanceValue(null);
+    setStatus(null);
+    setIsLoadingAllowance(false);
+    setIsProcessing(false);
+    setTransactionHash(null);
+  };
+
+  const onModalClose = !isPending ? resetStateAndCancel : undefined;
+
+  const onUpdateStatus = (status: TxnStatus, message?: string) => {
+    setStatus({ statusType: status, message: message });
+  };
+
+  const onContinue = async (values: FormValues) => {
+    setIsLoadingAllowance(true);
+    try {
+      if (!address || values.paymentMethod === "fiat") return;
+
+      const allowance = await getRetirementAllowance({
+        userAddress: address,
+        token: values.paymentMethod,
+      });
+
+      allowance &&
+        setAllowanceValue(
+          allowance[values.paymentMethod].retirementAggregatorV2
+        );
+      setInputValues(values);
+    } catch (e) {
+      console.error(e);
+      setErrorMessage(t`something went wrong loading the allowance`);
+    } finally {
+      setIsLoadingAllowance(false);
+    }
+  };
+
+  const getApprovalValue = (): string => {
+    if (!inputValues?.totalPrice) return "0";
+
+    const onePercent = utils
+      .parseUnits(
+        inputValues.totalPrice,
+        getTokenDecimals(inputValues.paymentMethod)
+      )
+      .div("100");
+    const val = safeAdd(
+      inputValues.totalPrice,
+      formatUnits(onePercent, getTokenDecimals(inputValues.paymentMethod))
+    );
+    return val;
+  };
+
+  // compare with total price including fees
+  const hasApproval = () => {
+    return (
+      !!allowanceValue &&
+      !!inputValues &&
+      Number(allowanceValue) >= Number(inputValues.totalPrice)
+    );
+  };
+
+  // approve with total price including fees
+  const handleApproval = async () => {
+    if (!provider || !inputValues) return;
+    try {
+      inputValues.paymentMethod !== "fiat" &&
+        (await approveTokenSpend({
+          tokenName: inputValues.paymentMethod,
+          spender: "retirementAggregatorV2",
+          signer: provider.getSigner(),
+          value: getApprovalValue(),
+          onStatus: onUpdateStatus,
+        }));
+    } catch (e) {
+      console.error(e);
+    }
+  };
+
+  const onMakeRetirement = async () => {
+    if (!provider || !inputValues) return;
+
+    try {
+      setIsProcessing(true);
+      const receipt = await retireCarbonTransaction({
+        provider,
+        address,
+        projectAddress: props.project.projectAddress,
+        paymentMethod: inputValues.paymentMethod,
+        maxAmountIn: getApprovalValue(),
+        retirementToken: props.price.name.toLowerCase() as PoolToken,
+        quantity: inputValues.quantity,
+        beneficiaryAddress: inputValues.beneficiaryAddress,
+        beneficiaryName: inputValues.beneficiaryName,
+        retirementMessage: inputValues.retirementMessage,
+        onStatus: onUpdateStatus,
+      });
+      receipt.transactionHash && setTransactionHash(receipt.transactionHash);
```

Comment:
We definitely want to link to the retirement page itself. The URL can be constructed like this:
`https://carbonmark.com/retirements/${address}/${retirementIndex}`

In order to do that we need to know the index of the retirement, which can be done like this:

https://github.com/KlimaDAO/klimadao/blob/dcb41e854249dd497d3de3f3c7bda581120f1a1e/examples/nextjs-typescript-integration/app/api/retire/route.ts#L75-L83

Once we have the URL, we can use Next `<Link>` or `useRouter().prefetch(url)` so that we can kickstart the page generation before they actually click the link.

Reply:
This did not work for me `staticCallResult is not a function` 

I used `getTotalRetirements` instead (easier)